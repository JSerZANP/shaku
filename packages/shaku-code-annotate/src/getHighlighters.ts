import { Highlighter, HighlighterOptions, getHighlighter } from "shiki";
import { codeToShakuHtml } from "./codeToShakuHtml";

interface ShakuHighlighter extends Highlighter {
  codeToShakuHtml: typeof codeToShakuHtml;
}

const shikiHighlighterFetcherStore = new Map<
  string,
  Promise<Array<ShakuHighlighter>>
>();

export async function getShakuHighlighters(options: HighlighterOptions) {
  const key = JSON.stringify(options);
  if (shikiHighlighterFetcherStore.has(key)) {
    return shikiHighlighterFetcherStore.get(key)!;
  }

  const themes = (options == null ? void 0 : options.themes) ?? [
    (options == null ? void 0 : options.theme) ?? "github-light",
  ];

  const highlighterFetcher = Promise.all(
    themes.map((theme) =>
      getHighlighter({
        ...(options ?? {}),
        theme,
        themes: void 0,
      }).then((highlighter) => {
        return Object.assign(highlighter, {
          codeToShakuHtml,
        });
      })
    )
  );
  shikiHighlighterFetcherStore.set(key, highlighterFetcher);
  return highlighterFetcher;
}
