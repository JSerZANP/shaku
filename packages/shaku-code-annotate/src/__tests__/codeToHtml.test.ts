import { expect, test } from "vitest";
import { getShakuHighlighters } from "../getHighlighters";

const snippets = [
  `
function ChatRoom({ roomId }) {
  const [serverUrl, setServerUrl] = useState('https://localhost:1234');

  useEffect(() => {
     // (                                                     )
        const connection = createConnection(serverUrl, roomId);
     // (1                )
        connection.connect();
    
  	return () => {
           //(2                    )
             connection.disconnect();
  	};
   //(3                 )
  }, [serverUrl, roomId]);
}
`,
];

test("codeToHtml()", async () => {
  const highlighters = await getShakuHighlighters({
    // @ts-ignore
    langs: ["js"],
  });

  snippets.forEach((code) => {
    const html = highlighters[0].codeToShakuHtml({
      code,
      meta: "annotate",
      parseBasicMarkdown: (code) => code,
      options: {},
    });

    expect(html).toMatchSnapshot();
  });
});
