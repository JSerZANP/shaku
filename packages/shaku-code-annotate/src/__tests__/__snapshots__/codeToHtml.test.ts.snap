// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codeToHtml() 1`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: undefined\\"></span></div><div class=\\"line\\"><span style=\\"color: undefined\\">function ChatRoom({ roomId }) {</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  const [serverUrl, setServerUrl] = useState('https://localhost:1234');</span></div><div class=\\"line\\"><span style=\\"color: undefined\\"></span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  useEffect(() =&gt; {</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">     // (                                                     )</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">        const connection = createConnection(serverUrl, roomId);</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">     // (1                )</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">        connection.connect();</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">    </span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  	return () =&gt; {</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">           //(2                    )</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">             connection.disconnect();</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  	};</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">   //(3                 )</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  }, [serverUrl, roomId]);</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">}</span></div><div class=\\"line\\"><span style=\\"color: undefined\\"></span></div></code></div></pre>",
  "skipped": false,
}
`;

exports[`codeToHtml() 2`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: undefined\\"></span></div><div class=\\"line\\"><span style=\\"color: undefined\\">function hello() {</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  //        ^!</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  //    [Hello world!]!</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  //        ^</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  //    [Shaku Shaku]</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  const blog = \\"https://jser.dev\\"</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">}</span></div><div class=\\"line\\"><span style=\\"color: undefined\\">  </span></div></code></div></pre>",
  "skipped": false,
}
`;
